### DISCUSSION TITLE
Using executables from multiple containers

### DISCUSSION THREAD

**Discussion by npavlovikj** [neurodesk/neurodesk.github.io-hugo#261] [Q&A] [answered]
Dear,

I have a question regarding using executables from multiple containers.

For example, I want to use the executable `@SUMA_Make_Spec_FS` from `afni`, that uses `mris_convert` from `freesurfer` underneath.
Based on this, [https://neurostars.org/t/mris-convert-within-fmriprep-docker/23026/5](https://neurostars.org/t/mris-convert-within-fmriprep-docker/23026/5), my understanding was that I can load multiple modules/containers and the executables from both should be available to each other.

However, that is not the case in my test:
```
(neurocommand) [npavlovikj@~ neurocommand]$ module load freesurfer/7.3.2
(neurocommand) [npavlovikj@~ neurocommand]$ module load afni/21.2.00 
(neurocommand) [npavlovikj@~ neurocommand]$ @SUMA_Make_Spec_FS -sid SK11 -fspath /work/npavlovikj/data_test/
++ Running @SUMA_Make_Spec_FS version: 2.2.4
program not found in path: mris_convert
1 program(s) not found
program failure: exiting...
```

`mris_convert` is part of the `freesurfer` module, and both loaded modules are part of my `$PATH` variable:
```
(neurocommand) [npavlovikj@~ neurocommand]$ which mris_convert
/lustre/work/npavlovikj/neurocommand/local/containers/freesurfer_7.3.2_20220812/mris_convert

(neurocommand) [npavlovikj@~ neurocommand]$ echo $PATH
/work/npavlovikj/neurocommand/local/containers/afni_21.2.00_20210714:/work/npavlovikj/neurocommand/local/containers/freesurfer_7.3.2_20220812:/home/npavlovikj/.conda/envs/neurocommand/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/util/opt/bin
```

Am I missing something in my setup to have this working?
I just cloned the `neurocommand` repo, so I am using the newest version available on `master`.

In the discussion here, [https://github.com/orgs/NeuroDesk/discussions/210](https://github.com/orgs/NeuroDesk/discussions/210), the `mris_convert` executable from `freesurfer` was added to a newer version of the `afni` container.
However, my understanding of NeuroDesk is that this can also be achieved by loading the individual `afni` and `freesurfer` modules?
Is this not necessarily true?

I would appreciate any feedback or suggestion you can provide.

Thank you!

--
Natasha

**Comment by stebo85** [comment, marked as answer]
Dear @npavlovikj 

Excellent question :)

The reason that mris_convert cannot be found in your example is the following:

Once you execute @SUMA_Make_Spec_FS 

then this is running inside the afni container and this script can only see things that are on the path inside the afni container. You are using 
the afni/21.2.00 container, which didn't have mris_convert included yet. If you switch to a later afni container version (e.g. 23.0.07) mris_convert is included and on the path inside the afni container, so your error will disappear :)

The module loading of executables will only work if binaries are not nested, e.g. you can call @SUMA_Make_Spec_FS and it will run inside afni and you can call mris_convert and it will run inside the freesurfer container. But the afni container (and therefore @SUMA_Make_Spec_FS) cannot talk to the freesurfer container - otherwise we would end up in an unreproducible setup.

Let us know if this works for you and makes sense :)

**Reply by npavlovikj** [reply]
Dear Steffen,

Thank you so much for the prompt reply and nice clarification - that all makes sense to me now.

Thank you,
Natasha
